File: NEvo/classes/Neuron.py
"""
Created on January 1 00:00 2025
@author: Johan Andreas Stendal
This class defines the neural evolution neuron
"""

from configNEvo import *

class Neuron:
    def __init__(self):
        self.Inputs = {}
        self.OutputWeights = {}
        self.networkNodeType = NetworkNodeType.HIDDEN
        self.layerNumber = 0
        self.output = 0.0

    def AddInputConnection(self, inputID : int):
        self.Inputs[inputID] = 0.0

    def AddOutputConnection(self, outputID : int, weight : float):
        self.OutputWeights[outputID] = weight

    def FeedInput(self, inputID : int, inputValue : float):
        self.Inputs[inputID] = inputValue

    def ActivationFunctionBoundedReLu(self, inputValue : float):
        return min(max(0.0, inputValue), 1.0)

    def ActivationFunctionReLu(self, inputValue : float):
        return max(0.0, inputValue)

    def ActivationFunctionSigmoid(self, inputValue : float):
        return 1.0 / (1.0 + math.exp(-inputValue))

    def CalculateOutput(self):
        inputSum = 0.0
        for inputValue in self.Inputs.values():
            inputSum += inputValue
        # only apply activation function if not input neuron
        if self.networkNodeType == NetworkNodeType.INPUT:
            self.output = inputSum
        else:
            self.output = self.ActivationFunctionBoundedReLu(inputSum)

    def SetNetworkNodeType(self, newNetworkNodeType : NetworkNodeType):
        self.networkNodeType = newNetworkNodeType

    def SetLayerNumber(self, newLayerNumber : int):
        self.layerNumber = newLayerNumber

    def GetInputs(self):
        return self.Inputs

    def GetOutputWeights(self):
        return self.OutputWeights

    def GetNetworkNodeType(self):
        return self.networkNodeType

    def GetLayerNumber(self):
        return self.layerNumber
        
    def GetOutputValue(self):
        return self.output

File: NeuralLife/NeuralLife/neuron.cpp
#pragma once
#include "neuron.hpp"

Neuron::Neuron()
{
	type = HIDDEN;
	layerNumber = 0;
	inputSum = 0.f;
	output = 0.f;
}

Neuron::~Neuron()
{

}

// add input connection with input ID (ID of neuron the input came from), input value set to 0 and activated to false
void Neuron::AddInputConnection(size_t inputID)
{
	Inputs.insert({ inputID, 0.f }); // inputID is the ID of the neuron giving the input
}

void Neuron::AddOutputConnection(size_t outputID, float weight)
{
	OutputWeights.insert({ outputID, weight });
}

// feed input with input ID (ID of neuron the input came from), input value and sets neuron input to activated
void Neuron::FeedInput(size_t inputID, float inputValue)
{
	Inputs[inputID] = inputValue; // inputID is the ID of the neuron giving the input
}

// trying bounded ReLu to output max 1
float Neuron::ActivationFunctionBoundedReLu(float input)
{
	return{ std::min(std::max(0.f, input), 1.f) }; // Bounded ReLu
}

float Neuron::ActivationFunctionReLu(float input)
{
	return{ std::max(0.f, input) }; // ReLu
}

float Neuron::ActivationFunctionSigmoid(float input)
{
	return{ 1.f / (1.f + std::exp(-input)) }; // Sigmoid
	//return{ std::tanh(input) }; // tanh
}

void Neuron::CalculateOutput()
{
	inputSum = 0.f;
	for (auto& input : Inputs) {
		inputSum += input.second;
	}
	// only apply activation function if not input neuron
	if (type == NetworkNodeType(INPUT)) {
		output = inputSum;
	}
	else {
		output = ActivationFunctionBoundedReLu(inputSum);
	}
}

void Neuron::SetType(NetworkNodeType newType)
{
	type = newType;
}

void Neuron::SetLayerNumber(size_t newLayerNumber)
{
	layerNumber = newLayerNumber;
}

std::map<size_t, float> Neuron::GetInputs()
{
	return { Inputs };
}

std::map<size_t, float> Neuron::GetOutputWeights()
{
	return { OutputWeights };
}

NetworkNodeType Neuron::GetType()
{
	return{ type };
}

size_t Neuron::GetLayerNumber()
{
	return{ layerNumber };
}

float Neuron::GetOutputValue()
{
	return{ output };
}